loki/transform/transform_parametrise.py

```diff
@@ -0,0 +1,307 @@
+"""
+from loki.expression import symbols as sym
+from loki import ir
+from loki.visitors import Transformer, FindNodes
+from loki.tools.util import is_iterable, as_tuple
+from loki.transform.transformation import Transformation
+from loki.transform.transform_inline import inline_constant_parameters
+
+
+__all__ = ['ParametriseTransformation']
+
+
+class ParametriseTransformation(Transformation):
+    """
+    Parametrise variables with provided values.
+
+    This transformation checks for each subroutine (defined as driver or entry point) the arguments to be parametrised
+    according to :attr:`dic2p` and passes this information down the calltree.
+
+    .. note::
+
+        A sanity run-time check will be inserted at each entry point to check consistency of the provided value
+        and argument value at this point!
+
+    .. warning::
+
+        The subroutine/call signature(s) may be altered as arguments are converted to local parameters or int literals.
+        Therefore, consistency must be ensured, meaning all parts of the code calling subroutines that are transformed
+        and all possibly differing names of variables at the entry points must be included, otherwise the resulting
+        code will not compile correctly!
+
+    E.g., use this class like this:
+
+    .. code-block:: python
+
+        def error_stop(**kwargs):
+            msg = kwargs.get("msg")
+            return ir.Intrinsic(text=f'error stop "{msg}"'),
+
+        dic2p = {'a': 12, 'b': 11}
+
+        transformation = ParametriseTransformation(dic2p=dic2p, disable=("ignore_this_func", "ignore_another_func"),
+                                                   abort_callback=error_stop, entry_points=("driver1", "driver2"))
+
+        scheduler.process(transformation=transformation)
+
+    Parameters
+    ----------
+    dic2p: dict
+        Dictionary of variable names and corresponding values to be parametrised.
+    disable: tuple
+        Tuple of subroutines not to be processed.
+    replace_by_value: bool
+        Replace variables entirely by value (default: `False`)
+    entry_points: None or tuple
+        Subroutine names to be used as entry points for parametrisation. Default `None` uses driver(s) as
+        entry points.
+    abort_callback:
+        Callback routine used for error on sanity check.
+        Available arguments via ``kwargs``:
+
+        * ``msg`` - predefined error message
+        * ``routine`` - the routine executing the sanity check
+        * ``var`` - the variable getting checked
+        * ``value`` - the value the variable should have (according to :attr:`dic2p`)
+    key : str
+        Access identifier/key for the ``item.trafo_data`` dictionary. Only necessary to provide if several of
+        these transformations are carried out in succession.
+    """
+
+    _key = "ParametriseTransformation"
+
+    def __init__(self, dic2p, disable=(), replace_by_value=False, entry_points=None, abort_callback=None, key=None):
```

Comment:
I think it's customary to use `None` rather than empty lists/tuples as default arguments, even if the initialization code becomes a bit more verbose that way. Rationale is that the default argument initializer is a single global instance and, while it isn't a problem with tuples as they are immutable, you can easily get yourself into trouble with innocent looking code like this:
```python
def some_routine(arg=[]):
    arg += ['a']
    return arg
```
which produces:
```
>>> some_routine()
['a']
>>> some_routine()
['a', 'a']
>>> some_routine()
['a', 'a', 'a']
```